File: GCE/todo-backend/app/storage.py
from app.models import Task, Tag, SubTask
from datetime import datetime, timedelta
from typing import Optional


class InMemoryStorage:
    def __init__(self):
        self.tasks: dict[str, Task] = {}
        self.tags: dict[str, Tag] = {}
        self._init_sample_data()

    def _init_sample_data(self):
        """Initialize with some sample data"""
        sample_tags = [
        ]
        for tag in sample_tags:
            self.tags[tag.id] = tag

        now = datetime.utcnow()
        sample_tasks = []
        for task in sample_tasks:
            self.tasks[task.id] = task

    def get_all_tasks(
        self,
        status: str = "all",
        archived: bool = False,
        tags: Optional[list[str]] = None,
        page: int = 1,
        limit: int = 50,
    ) -> tuple[list[Task], int]:
        filtered_tasks = list(self.tasks.values())

        filtered_tasks = [t for t in filtered_tasks if t.archived == archived]

        if status == "active":
            filtered_tasks = [t for t in filtered_tasks if not t.completed]
        elif status == "completed":
            filtered_tasks = [t for t in filtered_tasks if t.completed]

        if tags:
            filtered_tasks = [
                t for t in filtered_tasks if any(tag in t.tags for tag in tags)
            ]

        filtered_tasks.sort(key=lambda t: t.createdAt, reverse=True)

        total = len(filtered_tasks)

        start = (page - 1) * limit
        end = start + limit
        paginated_tasks = filtered_tasks[start:end]

        return paginated_tasks, total

    def get_task(self, task_id: str) -> Optional[Task]:
        return self.tasks.get(task_id)

    def create_task(self, task: Task) -> Task:
        self.tasks[task.id] = task
        return task

    def update_task(self, task_id: str, task: Task) -> Optional[Task]:
        if task_id in self.tasks:
            self.tasks[task_id] = task
            return task
        return None

    def delete_task(self, task_id: str) -> bool:
        if task_id in self.tasks:
            del self.tasks[task_id]
            return True
        return False

    def get_all_tags(self) -> list[Tag]:
        return list(self.tags.values())

    def get_tag(self, tag_id: str) -> Optional[Tag]:
        return self.tags.get(tag_id)

    def create_tag(self, tag: Tag) -> Tag:
        self.tags[tag.id] = tag
        return tag

    def update_tag(self, tag_id: str, tag: Tag) -> Optional[Tag]:
        if tag_id in self.tags:
            self.tags[tag_id] = tag
            return tag
        return None

    def delete_tag(self, tag_id: str) -> bool:
        if tag_id in self.tags:
            del self.tags[tag_id]
            return True
        return False

    def get_statistics(self) -> dict:
        tasks = list(self.tasks.values())
        return {
            "total": len(tasks),
            "completed": len([t for t in tasks if t.completed and not t.archived]),
            "active": len([t for t in tasks if not t.completed and not t.archived]),
            "advanced": len([t for t in tasks if t.type == "advanced"]),
            "archived": len([t for t in tasks if t.archived]),
        }

    def get_contributions(
        self, start_date: Optional[str] = None, end_date: Optional[str] = None, days: int = 365
    ) -> tuple[list[dict], dict]:
        today = datetime.utcnow().date()

        if end_date:
            end = datetime.strptime(end_date, "%Y-%m-%d").date()
        else:
            end = today

        if start_date:
            start = datetime.strptime(start_date, "%Y-%m-%d").date()
        else:
            start = end - timedelta(days=days - 1)

        contributions = {}
        for task in self.tasks.values():
            if task.completedAt:
                date_str = task.completedAt.strftime("%Y-%m-%d")
                task_date = task.completedAt.date()
                if start <= task_date <= end:
                    contributions[date_str] = contributions.get(date_str, 0) + 1

        contribution_list = []
        current = end
        while current >= start:
            date_str = current.strftime("%Y-%m-%d")
            contribution_list.append({
                "date": date_str,
                "count": contributions.get(date_str, 0)
            })
            current -= timedelta(days=1)

        total_contributions = sum(c["count"] for c in contribution_list)
        
        current_streak = 0
        longest_streak = 0
        temp_streak = 0
        
        for contrib in contribution_list:
            if contrib["count"] > 0:
                temp_streak += 1
                longest_streak = max(longest_streak, temp_streak)
            else:
                temp_streak = 0

        for contrib in contribution_list:
            if contrib["count"] > 0:
                current_streak += 1
            else:
                break

        summary = {
            "totalContributions": total_contributions,
            "longestStreak": longest_streak,
            "currentStreak": current_streak,
        }

        return contribution_list, summary


storage = InMemoryStorage()

File: GCE/todo-backend/src/storage.js
// In-memory storage
const storage = {
  tasks: new Map(),
  tags: new Map(),
};

// Task methods
export function getAllTasks({
  status = "all",
  archived = false,
  tags = null,
  page = 1,
  limit = 50,
}) {
  let tasks = Array.from(storage.tasks.values());

  // Filter by archived
  tasks = tasks.filter((t) => t.archived === archived);

  // Filter by status
  if (status === "active") {
    tasks = tasks.filter((t) => !t.completed);
  } else if (status === "completed") {
    tasks = tasks.filter((t) => t.completed);
  }

  // Filter by tags
  if (tags && tags.length > 0) {
    tasks = tasks.filter((t) => tags.some((tag) => t.tags.includes(tag)));
  }

  // Sort by createdAt descending
  tasks.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));

  const total = tasks.length;
  const start = (page - 1) * limit;
  const paginatedTasks = tasks.slice(start, start + limit);

  return { tasks: paginatedTasks, total };
}

export function getTask(taskId) {
  return storage.tasks.get(taskId) || null;
}

export function createTask(task) {
  storage.tasks.set(task.id, task);
  return task;
}

export function updateTask(taskId, task) {
  if (storage.tasks.has(taskId)) {
    storage.tasks.set(taskId, task);
    return task;
  }
  return null;
}

export function deleteTask(taskId) {
  return storage.tasks.delete(taskId);
}

// Tag methods
export function getAllTags() {
  return Array.from(storage.tags.values());
}

export function getTag(tagId) {
  return storage.tags.get(tagId) || null;
}

export function createTag(tag) {
  storage.tags.set(tag.id, tag);
  return tag;
}

export function updateTag(tagId, tag) {
  if (storage.tags.has(tagId)) {
    storage.tags.set(tagId, tag);
    return tag;
  }
  return null;
}

export function deleteTag(tagId) {
  return storage.tags.delete(tagId);
}

// Statistics
export function getStatistics() {
  const tasks = Array.from(storage.tasks.values());
  return {
    total: tasks.length,
    completed: tasks.filter((t) => t.completed && !t.archived).length,
    active: tasks.filter((t) => !t.completed && !t.archived).length,
    advanced: tasks.filter((t) => t.type === "advanced").length,
    archived: tasks.filter((t) => t.archived).length,
  };
}

export function getContributions({
  startDate = null,
  endDate = null,
  days = 365,
}) {
  const today = new Date();
  today.setHours(0, 0, 0, 0);

  const end = endDate ? new Date(endDate) : today;
  const start = startDate
    ? new Date(startDate)
    : new Date(end.getTime() - (days - 1) * 24 * 60 * 60 * 1000);

  const contributions = {};
  const tasks = Array.from(storage.tasks.values());

  for (const task of tasks) {
    if (task.completedAt) {
      const taskDate = new Date(task.completedAt);
      taskDate.setHours(0, 0, 0, 0);
      if (taskDate >= start && taskDate <= end) {
        const dateStr = taskDate.toISOString().split("T")[0];
        contributions[dateStr] = (contributions[dateStr] || 0) + 1;
      }
    }
  }

  const contributionList = [];
  const current = new Date(end);
  while (current >= start) {
    const dateStr = current.toISOString().split("T")[0];
    contributionList.push({
      date: dateStr,
      count: contributions[dateStr] || 0,
    });
    current.setDate(current.getDate() - 1);
  }

  const totalContributions = contributionList.reduce(
    (sum, c) => sum + c.count,
    0,
  );

  let currentStreak = 0;
  let longestStreak = 0;
  let tempStreak = 0;

  for (const contrib of contributionList) {
    if (contrib.count > 0) {
      tempStreak++;
      longestStreak = Math.max(longestStreak, tempStreak);
    } else {
      if (currentStreak === 0 && tempStreak > 0) {
        currentStreak = tempStreak;
      }
      tempStreak = 0;
    }
  }

  if (contributionList[0]?.count > 0) {
    currentStreak = tempStreak > 0 ? tempStreak : currentStreak;
  }

  return {
    data: contributionList,
    summary: { totalContributions, longestStreak, currentStreak },
  };
}
